src/tree/gpu_hist/histogram.cu

```diff
@@ -0,0 +1,156 @@
+#include <thrust/sort.h>
+#include <thrust/iterator/discard_iterator.h>
+#include "xgboost/base.h"
+#include "histogram.cuh"
+#include "row_partitioner.cuh"
+#include "../../data/ellpack_page.cuh"
+#include "../../common/device_helpers.cuh"
+#include "../updater_gpu_common.cuh"
+
+namespace xgboost {
+namespace tree {
+// Following 2 functions are slightly modifed version of fbcuda.
+
+/* \brief Constructs a rounding factor used to truncate elements in a sum such that the
+   sum of the truncated elements is the same no matter what the order of the sum is.
+
+ * Follows Algorithm 5: Reproducible Sequential Sum in 'Fast Reproducible Floating-Point
+ * Summation' by Demmel and Nguyen
+
+ * @param max max(sum(positive gradient) + sum(abs(negative gradient)))
+ * @param n The number of elements for the sum, or an upper bound estimate
+ */
+DEV_INLINE __host__ float CreateRoundingFactor(float max, int n) {
```

Comment:
I was expecting you to do this with doubles, this would give an extra 29 bits of precision and might resolve your accuracy issues. We can make the check to only do the truncation if GradientSumT is double precision.